### PR TITLE
Message: add to_pandas wrapper method and .write as alias for compat

### PR DESCRIPTION
**Pull request by dr-leo**
Extended writer tests. But this adds a log of duplication. Maybe check wrappers separately for efficiency.

### CODE REVIEW

**Comment by dr-leo** [comment]
This mitigates some backwards compat issues discussed in #145. In addition, it makes conversion to pandas objects more convenient as the to_pandas function needs not be imported.

File: pandasdmx/message.py [hunk 1]
```diff
@@ -92,6 +93,26 @@ class Config:
...
+        :meth:`Message.write`. 
+        It is deprecated and may be removed in future versions.
+        '''
+        from pandasdmx.writer import write
+        return write(self, *args, **kwargs)
+    
+    def write(self, *args, **kwargs):
+        'Deprecated alias for :meth:`Message.to_pandas`'
+        warn('Message.write() is deprecated. Use Message.to_pandas() instead.',
+            DeprecationWarning)
```
Comment (khaeru) [inline pandasdmx/message.py:113]: Check indentation.

File: pandasdmx/message.py [hunk 2]
```diff
@@ -103,7 +124,7 @@ def __repr__(self):
         ]
         lines.extend(_summarize(self, ['footer', 'response']))
         return '\n  '.join(lines)
-
+
```
Comment (khaeru) [inline pandasdmx/message.py:127]: Delete whitespace on empty lines.

File: pandasdmx/tests/test_writer.py [hunk 3]
```diff
@@ -54,21 +54,28 @@ def test_write_data_arguments():
...
         sdmx.to_pandas(msg, attributes='foobarbaz')
+        msg.to_pandas(attributes='foobarbaz')
 
 
 def test_write_data(data_path):
     msg = sdmx.read_sdmx(data_path)
 
     result = sdmx.to_pandas(msg)
+    # Check if wrapper method does the same.
+    # This involves a lot of duplication. Maybe test only for one case.
```
Comment (khaeru) [inline pandasdmx/tests/test_writer.py:71]: Agreed!

The convenience method contains no code, and only chains to the actual writer method. It is not necessary to test its return value; only that it (a) runs without exception and (b) returns a pandas object.

Instead expanding a parametrized test like this one, remove the lines, and either add to or duplicate (with appropriate modifications and comments) one of the tests further down in the file that handles a *single* message, e.g. `test_write_categoryscheme`.

File: pandasdmx/message.py [hunk 4]
```diff
@@ -92,6 +93,26 @@ class Config:
     #: that returned the Message. This is not part of the SDMX standard.
     response: Response = None
 
+    def to_pandas(self, *args, **kwargs):
+        '''
+        Convert a :inst:`Message` to :mod:`pandas` object(s) by
```
Comment (khaeru) [inline pandasdmx/message.py:98]: - `:inst:` is not a [Sphinx ReST role](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html). 
- Within a class's docstring or the docstring of its methods, it is not necessary to use any special markup to refer to the class. See below where 'DataMessage' is used without markup in the docstring of that class.

File: pandasdmx/message.py [hunk 5]
```diff
@@ -92,6 +93,26 @@ class Config:
...
+        calling :func:`pandasdmx.writer.write`. 
+        The :inst:`Message` is passed implicitly.
```
Comment (khaeru) [inline pandasdmx/message.py:100]: This is always true of a class method; remove.

File: pandasdmx/message.py [hunk 6]
```diff
@@ -92,6 +93,26 @@ class Config:
...
+        
+        For  backwards compat with previous releases :class:`Message` has an identical  alias
+        :meth:`Message.write`. 
+        It is deprecated and may be removed in future versions.
+        '''
+        from pandasdmx.writer import write
+        return write(self, *args, **kwargs)
+    
+    def write(self, *args, **kwargs):
+        'Deprecated alias for :meth:`Message.to_pandas`'
```
Comment (khaeru) [inline pandasdmx/message.py:111]: Simply call it "Alias for…", then use a [deprecated ReST directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated).

**Comment by dr-leo** [comment]
This is not quite ready for merging. For some obscure reason one test fails due to reversed dim order issue. I thought this was fixed. - My refactoring of Request._request_from_args should not have changed the semantics.  I was struck by a bug in this method where the dsd was not popped from the args as it should have. This bug should have surfaced before though. It broke compilation of walkthrough.rst.

I am at a loss now.

**Comment by khaeru** [comment]
> I was struck by a bug in this method where the dsd was not popped from the args as it should have. This bug should have surfaced before though. It broke compilation of walkthrough.rst.

What was the symptom?

> I am at a loss now.

Would you like me to fix the code?

**Comment by dr-leo** [comment]
Let me try to explain why I am at a loss. Consider the following pytest and pdb output:

```
pandasdmx\tests\test_docs.py F
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured stdout >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
CL_AGE
Y15-24    From 15 to 24 years
Y15-74    From 15 to 74 years
Y20-64    From 20 to 64 years
Y25-54    From 25 to 54 years
Y25-74    From 25 to 74 years
Y55-74    From 55 to 74 years
Name: AGE, dtype: object
CL_UNIT
THS_PER                   Thousand persons
PC_POP      Percentage of total population
PC_ACT     Percentage of active population
Name: UNIT, dtype: object
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

    @pytest.mark.remotedata
    def test_doc_example():
        """Code from example.rst."""
        import pandasdmx as sdmx
        estat = sdmx.Request('ESTAT')

        metadata = estat.datastructure('DSD_une_rt_a')

        for cl in 'CL_AGE', 'CL_UNIT':
            print(sdmx.to_pandas(metadata.codelist[cl]))

        resp = estat.data(
            'une_rt_a',
            key={'GEO': 'EL+ES+IE'},
            params={'startPeriod': '2007'},
            )

        data = sdmx.to_pandas(resp) \
                   .xs('Y15-74', level='AGE', drop_level=False)

        data.loc[('PC_ACT', 'Y15-74', 'T')]

        # Further checks per https://github.com/dr-leo/pandaSDMX/issues/157

        # DimensionDescriptor for the structure message
        dd1 = metadata.structure.DSD_une_rt_a.dimensions

        # DimensionDescriptor retrieved whilst validating the data message
        dd2 = resp.data[0].structured_by.dimensions

        # DimensionDescriptors have same ID, components and order
>       assert dd1 == dd2
E       assert <DimensionDes... TIME_PERIOD>> == <DimensionDes...ension: FREQ>>
E         Use -v to get the full diff

pandasdmx\tests\test_docs.py:54: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> c:\users\a3305\documents\github\pandasdmx\pandasdmx\tests\test_docs.py(54)test_doc_example()
-> assert dd1 == dd2
(Pdb) dd1
<DimensionDescriptor: <Dimension: FREQ>; <Dimension: AGE>; <Dimension: UNIT>; <Dimension: SEX>; <Dimension: GEO>; <TimeDimension: TIME_PERIOD>>
(Pdb)
<DimensionDescriptor: <Dimension: FREQ>; <Dimension: AGE>; <Dimension: UNIT>; <Dimension: SEX>; <Dimension: GEO>; <TimeDimension: TIME_PERIOD>>
(Pdb) dd1
<DimensionDescriptor: <Dimension: FREQ>; <Dimension: AGE>; <Dimension: UNIT>; <Dimension: SEX>; <Dimension: GEO>; <TimeDimension: TIME_PERIOD>>
(Pdb) dd1[0]
<Dimension: FREQ>
(Pdb) dd1[0].order
1
(Pdb) dd1
<DimensionDescriptor: <Dimension: FREQ>; <Dimension: AGE>; <Dimension: UNIT>; <Dimension: SEX>; <Dimension: GEO>; <TimeDimension: TIME_PERIOD>>
(Pdb) dd1[1].order
2
(Pdb) dd2
<DimensionDescriptor: <Dimension: TIME_PERIOD>; <Dimension: UNIT>; <Dimension: AGE>; <Dimension: SEX>; <Dimension: GEO>; <Dimension: FREQ>>
(Pdb) data
UNIT     AGE     SEX  GEO  FREQ  TIME_PERIOD
PC_ACT   Y15-74  F    EL   A     2007            12.9
                                 2008            11.5
                                 2009            13.3
     

     type(data)
<class 'pandas.core.series.Series'>
(Pdb) data.index
MultiIndex([( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2007'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2008'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2009'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2010'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2011'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2012'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2013'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2014'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2015'),
            ( 'PC_ACT', 'Y15-74', 'F', 'EL', 'A', '2016'),
            ...
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2010'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2011'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2012'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2013'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2014'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2015'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2016'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2017'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2018'),
            ('THS_PER', 'Y15-74', 'T', 'IE', 'A', '2019')],
           names=['UNIT', 'AGE', 'SEX', 'GEO', 'FREQ', 'TIME_PERIOD'], length=351)
```

Some observations:
* dim order in dd2 is still reversed. The order in dd1 is correct. This should have been fixed in #161.
* In #161, you edited doc/example.py. Therein, `data`is rendered as pd.Series rather than pd.DataFrame as it should by default since #151. While you rightly reminded me to check whether the docs build properly, I wonder if they did after #151 and even #163. It seems to me as if the example snippet did not reflect that `data` is rendered as dataframe now.
* In api.Request._request_from_args there is the line `dsd = kwargs.get('dsd', None) (I believe line 204). The subsequent check for invalid kwargs then fails. And further down, the dsd item is re-added to the kwargs as if it had been popped. I streamlined all this code in the current commit (the only difference being that invalid kwargs raise TypeError rather than ValueError).

I will try to track this further down within the next few days. To do this, I will have to research how to update the working dir to an older commit in Github Desktop.

**Comment by khaeru** [comment]
> - dim order in dd2 is still reversed. The order in dd1 is correct. This should have been fixed in #161.

You mean #163, which indeed fixed #157. The 'streamlining' (per the third bullet) on this branch breaks this functionality.

Again, I'll offer to resolve the issue by adjusting the branch to remove unnecessary changes that break the functionality you previously asked for.

> - While you rightly reminded me to check whether the docs build properly, I wonder if they did after #151 and even #163.

The ReadTheDocs hook is still broken and still needs you, as admin of the repository, to fix it. That would make it possible to check that the docs build correctly after every commit. As it is, I build them locally.

**Comment by dr-leo** [comment]
I've added the WebHook on readthedocs Any other admin Tasks in this context?

Frankly, until 2018 or so RTD built the docs on each commit to specified branches. So I am surprised that it no longer did.But if this feature was removed/replaced with the current WebHook, it should be ok now.

**Comment by dr-leo** [comment]
I consider this to be ready for merging. The build errors on MacOS for other reasons.

However, #163 requires additional work. It should not have been merged in the first place. It is wrong to use the DSD to determin the dim order for a generic dataset. Rather, the DSD is only needed for structure-specific datasets. As I suspected in the description of the second-last commit, dim order is still reversed when a str-typed key is passed to construct a data request.

**Pull request merged**